src/hotspot/share/oops/compressedKlass.hpp

```diff
@@ -66,11 +66,29 @@ class CompressedKlassPointers : public AllStatic {
   friend class VMStructs;
   friend class ArchiveBuilder;
 
-  // Encoding base
-  static address _base;
-
-  // Shift is actually a constant; we keep this just for the SA (see vmStructs.cpp and
+  // A dense representation of values one often loads in quick succession, in order to fold
+  // all of them into a single load:
+  // - UseCompactObjectHeaders and UseCompressedClassPointers flags
+  // - encoding base and encoding shift
+  // We can encode everything (including the encoding base) in a 64-bit word. The encoding
+  // base will always be page aligned, so we have a 12-bit alignment shadow to store the rest
+  // of the data.
+
+  static uintptr_t _value;
+  static constexpr int useCompactObjectHeadersShift = 0;
+  static constexpr int useCompressedClassPointersShift = 1;
+  static constexpr int encodingShiftShift = 2;
+  static constexpr int encodingShiftWidth = 5;
+  static constexpr int baseAddressMask = ~right_n_bits(12);
+
+  static void set_encoding_base(address base);
+  static void set_encoding_shift(int shift);
+  static void set_use_compact_headers(bool b);
+  static void set_use_compressed_class_pointers(bool b);
+
+  // These members hold copies of encoding base and shift and only exist for SA (see vmStructs.cpp and
```

Comment:
I guess, SA could be changed to decode the config bits? Maybe not in this PR, though.

Reply:
Yes, we should fix that.